### PR TITLE
verbosify cirq_type deprecation

### LINKED ISSUE #5337: It's too late to deprecate others using `cirq_type` in their json
There's a bad deprecation warning in the JSON serialization:

```
def _json_dict_with_cirq_type(obj: Any):
    base_dict = obj._json_dict_()
    if 'cirq_type' in base_dict:
        # TODO: upgrade to ValueError in v0.15
        warnings.warn(
            f"Found 'cirq_type': '{base_dict['cirq_type']}' in _json_dict_. "
            f"Custom values of this field are not permitted, and will produce "
            "an error starting in Cirq v0.15.",
            DeprecationWarning,
        )
        return base_dict
    return {'cirq_type': json_cirq_type(type(obj)), **base_dict}
```

**This cannot be changed into an error**. The main guarantee provided by json serialization is that it will continue to work forever. If using this key was going to be an error, it had to start as an error when json serialization was first written. You can't change it now, because that might break things which would mean they wouldn't continue to work forever.

I am aware of specific files out there that use these values, because stimcirq uses these keys for its custom operations. I've serialized circuits that had those operations in it. I expect them to continue to parse.

You are about to make a change that would break the main guarantee provided by json serialization. DO NOT DO THIS. Remove this deprecation warning instead of moving forward with it.

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Fixes #5337.

More specifically, this message is intended to make it very clear that:

1. `cirq_type` will persist as a field in JSON-serialized objects
2. `cirq_type` is now auto-generated
3. Users should remove `cirq_type` from their `_json_dict_` methods
4. Classes outside Cirq should define `_json_namespace_` to avoid collisions

### CODE REVIEW

File: cirq-core/cirq/protocols/json_serialization.py [hunk 1]
```diff
@@ -283,9 +283,14 @@ def _json_dict_with_cirq_type(obj: Any):
...
         warnings.warn(
-            f"Found 'cirq_type': '{base_dict['cirq_type']}' in _json_dict_. "
-            f"Custom values of this field are not permitted, and will produce "
-            "an error starting in Cirq v0.15.",
+            f"Found 'cirq_type': '{base_dict['cirq_type']}' in user-specified _json_dict_. "
+            "This field is now automatically generated from the class name and (if present) "
+            "its _json_namespace_ method. Starting in v0.15, custom 'cirq_type' values "
+            "will trigger an error."
+            "\n\n"
+            "To fix this, remove 'cirq_type' from the class _json_dict_ method. Classes "
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/json_serialization.py:291]: Should we also mention that serialized objects will continue to be the same as previously serialized objects?

**Review by tanujkhattar — APPROVED** [review]

File: cirq-core/cirq/protocols/json_serialization.py (hunk 1 above)
Comment (95-martin-orion) [reply cirq-core/cirq/protocols/json_serialization.py:292]: Emphasized this in the preceding paragraph. FWIW, this only preserves serialization format if a class follows the `[namespace.]classname` convention, but we've been _exceedingly_ consistent about applying that convention in Cirq. Hopefully that translates to external users doing the same.

**Comment by 95-martin-orion** [comment]
@Strilanc, I'd still like your confirmation on this before merging, since you were the one affected by the previous wording here.

**Comment by Strilanc** [comment]
Shouldn't 'how to fix this' vary depending on whether the manually assigned value agrees with the automatically assigned value or not? If I was using a custom cirq_type value, how would I override the automatic behavior? How would I guarantee cross-compatibility across versions with/without this deprecation?

**Comment by 95-martin-orion** [comment]
Okay, I think I hit the points raised above.

> If I was using a custom cirq_type value, how would I override the automatic behavior?

You can't - the point of this change was to enforce consistency in `cirq_type`s. The motivation for this decision and discussion around it can be found on PR #4693.

> How would I guarantee cross-compatibility across versions with/without this deprecation?

As described in #5337, you'd need to include both `cirq_type`s in the deserialization dict. This means that external classes which opted not to use a namespace will never be totally safe from the risk of colliding with a future Cirq class name, but that was true before this change as well.

LMK if there's anything else you'd like to see covered here.

File: cirq-core/cirq/protocols/json_serialization.py [hunk 2]
```diff
@@ -284,14 +284,16 @@ def _json_dict_with_cirq_type(obj: Any):
...
             "\n\n"
             "Starting in v0.15, custom 'cirq_type' values will trigger an error. "
-            "To fix this, remove 'cirq_type' from the class _json_dict_ method. Classes "
-            "outside of cirq should also define a non-empty _json_namespace_ to avoid "
-            "collision with Cirq classes.",
+            "To fix this, remove 'cirq_type' from the class _json_dict_ method and "
+            "define _json_namespace_ for the class."
+            "\n\n"
+            "For backwards compatibility, classes whose old 'cirq_type' does not match the "
+            "new value must appear in `json_resolver_cache.py::_class_resolver_dictionary()` "
```
Comment (Strilanc) [inline cirq-core/cirq/protocols/json_serialization.py:295]: I don't understand this instruction. My project does not have a `json_resolver_cache.py` file. Is this a file within cirq? Say it is within cirq, and give the full file path not the name.

Also, this seems rather extreme to require someone who wrote a different library to make a change in cirq to maintaing compatibility across versions? It sounds like this is saying I *can't* maintain compatibility.

**Review by Strilanc — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/protocols/json_serialization.py (hunk 2 above)
Comment (95-martin-orion) [reply cirq-core/cirq/protocols/json_serialization.py:295]: Sorry, I misunderstood how Cirq handles third-party JSON resolvers. The appropriate thing to update for this is whatever custom resolver you are using, as described [here](https://github.com/quantumlib/Cirq/blob/efd3b25e83a6dacde69d7d000b7de19d8671cea8/cirq-core/cirq/protocols/json_serialization.py#L73). I'll update the instructions to match.

**Review by Strilanc — APPROVED** [review]

**Pull request merged**